insonmnia/oracle/oracle.go

```diff
@@ -0,0 +1,179 @@
+package oracle
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/noxiouz/zapctx/ctxlog"
+	"github.com/sonm-io/core/blockchain"
+	"github.com/sonm-io/core/util"
+	"go.uber.org/zap"
+)
+
+type Oracle struct {
+	key *ecdsa.PrivateKey
+	cfg *Config
+
+	logger *zap.Logger
+
+	bch blockchain.API
+
+	actualPrice  *big.Int
+	currentPrice *big.Int
+}
+
+func NewOracle(ctx context.Context, key *ecdsa.PrivateKey, cfg *Config) (*Oracle, error) {
+	logger := ctxlog.GetLogger(ctx)
+
+	var mode string
+	if cfg.Oracle.Mode {
+		mode = "master"
+	} else {
+		mode = "slave"
+	}
+	logger.Info("start USD-SNM Oracle",
+		zap.String("mode", mode),
+		zap.String("account", crypto.PubkeyToAddress(key.PublicKey).String()),
+		zap.String("price update period:", cfg.Oracle.PriceUpdatePeriod.String()),
+		zap.String("contract update period", cfg.Oracle.ContractUpdatePeriod.String()),
+		zap.Float64("deviation percent", cfg.Oracle.Percent))
+
+	bch, err := blockchain.NewAPI(blockchain.WithConfig(cfg.Blockchain))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Oracle{
+		logger:       logger,
+		key:          key,
+		cfg:          cfg,
+		bch:          bch,
+		currentPrice: big.NewInt(0),
+		actualPrice:  big.NewInt(0),
+	}, nil
+}
+
+func (o *Oracle) Serve(ctx context.Context) error {
+	priceWatcher := NewPriceWatcher(o.cfg.Oracle.PriceUpdatePeriod)
+	dw := priceWatcher.Start(ctx)
+
+	t := util.NewImmediateTicker(o.cfg.Oracle.ContractUpdatePeriod)
+
+	events, err := o.bch.Events().GetEvents(ctx, blockchain.GetMultiSigFilter([]common.Address{blockchain.OracleUsdMultisigAddr()}, big.NewInt(0)))
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case p := <-dw:
+			if p.err != nil {
+				return p.err
+			}
+			o.actualPrice = p.price
+			o.logger.Debug("loaded new price", zap.String("price", p.price.String()))
+		case <-t.C:
+			currentPrice, err := o.bch.OracleUSD().GetCurrentPrice(ctx)
+			if err != nil {
+				return err
+			}
+			if o.currentPrice.Cmp(currentPrice) != 0 {
```

Comment:
what if `<-dw` select file **after** the `t.C`? Price will be empty and everything will be fucked. So explicitly load price for first time, then start timer handling.